src/server/zset_family.cc

```diff
@@ -2599,6 +2646,301 @@ void ZSetFamily::GeoDist(CmdArgList args, ConnectionContext* cntx) {
+    if (HASHISZERO(neighbors[i])) {
+      continue;
+    }
+
+    // When a huge Radius (in the 5000 km range or more) is used,
+    // adjacent neighbors can be the same, leading to duplicated
+    // elements. Skip every range which is the same as the one
+    // processed previously.
+    if (last_processed && neighbors[i].bits == neighbors[last_processed].bits &&
+        neighbors[i].step == neighbors[last_processed].step) {
+      continue;
+    }
+
+    GeoHashFix52Bits min, max;
+    scoresOfGeoHashBox(neighbors[i], &min, &max);
+
+    ZSetFamily::ScoreInterval si;
+    si.first = ZSetFamily::Bound{static_cast<double>(min), false};
+    si.second = ZSetFamily::Bound{static_cast<double>(max), true};
+
+    ZSetFamily::RangeParams range_params;
+    range_params.interval_type = ZSetFamily::RangeParams::IntervalType::SCORE;
+    range_params.with_scores = true;
+    range_specs.emplace_back(si, range_params);
+
+    last_processed = i;
+  }
+
+  // get all the matching members and add them to the potential result list
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpRanges(range_specs, t->GetOpArgs(shard), key);
+  };
+  OpResult<vector<ScoredArray>> result_arrays =
+      cntx->transaction->ScheduleSingleHopT(std::move(cb));
+  if (result_arrays.status() == OpStatus::WRONG_TYPE) {
+    (*cntx)->SendError(kWrongTypeErr);
+    return -1;
+  }
+
+  // filter potential result list
+  unsigned long count = 0;
+  double xy[2];
+  double distance;
+  for (auto& arr : *result_arrays) {
+    for (auto& p : arr) {
+      if (geoWithinShape(shape, p.second, xy, &distance) == 0) {
+        ga->emplace_back(xy[0], xy[1], distance, p.second, p.first);
+        count++;
+        if (limit > 0 && ga->size() >= limit)
+          break;
+      }
+    }
+  }
+
+  return count;
+}
+
+void GeoSearchGeneric(ConnectionContext* cntx, const GeoShape& shape_ref, string_view key,
+                      const GeoSearchOpts& geo_ops) {
+  double conversion = geo_ops.conversion;
+  uint64_t count = geo_ops.count;
+  bool asc = geo_ops.asc;
+  bool desc = geo_ops.desc;
+  bool any = geo_ops.any;
+  bool withdist = geo_ops.withdist;
+  bool withcoord = geo_ops.withcoord;
+  bool withhash = geo_ops.withhash;
+
+  // query
+  GeoShape* shape = &(const_cast<GeoShape&>(shape_ref));
+  GeoHashRadius georadius = geohashCalculateAreasByShapeWGS84(shape);
+  GeoArray ga;
+  MembersOfAllNeighbors(cntx, key, georadius, shape_ref, &ga, any ? count : 0);
+
+  // if no matching results, the user gets an empty reply.
+  if (ga.empty()) {
+    (*cntx)->SendNull();
+  }
+
+  // sort and trim by count
+  if (asc) {
+    auto asc_comparator = [](const GeoPoint& a, const GeoPoint& b) { return a.dist < b.dist; };
+    if (count > 0) {
+      std::partial_sort(ga.begin(), ga.begin() + count, ga.end(), asc_comparator);
+      ga.resize(count);
+    } else {
+      std::sort(ga.begin(), ga.end(), asc_comparator);
+    }
```

Comment:
Can you create a function (in the anonymous namespace) with this, which accepts a comparator (probably of type `absl::FunctionRef<bool(const GeoPoint&, const GeoPoint&)>`)?
This way you both shorten this function, and remove code copy-paste

An alternative would be to create a function which will take the above-mentioned `enum` and sort internally. Something like:

```cpp
void SortIfNeeded(GeoArray* ga, Sorting sorting) {
  if (sorting == kUnsorted)
    return;

  auto comparator = [&](const GeoPoint& a, const GeoPoint& b) {
    if (sorting == Sorting::kAsc) {
      return a.dist < b.dist;
    } else {
      DCHECK(sorting == Sorting::kDesc);
      return a.dist > b.dist;
    }
  };

  if (count > 0) {
    std::partial_sort(ga.begin(), ga.begin() + count, ga.end(), comparator);
    ga.resize(count);
  } else {
    std::sort(ga.begin(), ga.end(), comparator);
  }
}
```

Reply:
I created `SortIfNeeded` as you suggested:)